source/common/stats/deferred_creation.h

```diff
@@ -0,0 +1,113 @@
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  DeferredCreation(const typename StatsStructType::StatNameType& stat_names,
+                   Stats::ScopeSharedPtr scope)
+      : initialized_([&scope]() -> Gauge& {
+          Stats::StatNamePool pool(scope->symbolTable());
+          return Stats::Utility::gaugeFromElements(
+              *scope, {pool.add(StatsStructType::typeName()), pool.add("initialized")},
+              Stats::Gauge::ImportMode::HiddenAccumulate);
+        }()),
+        ctor_([stats_scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          initialized_.inc();
+          // Reset ctor_ to save some RAM.
+          Cleanup reset_ctor([&] { ctor_ = nullptr; });
+          return new StatsStructType(stat_names, *stats_scope);
+        }) {
+    if (initialized_.value() > 0) {
+      instantiate();
+    }
+  }
+  ~DeferredCreation() {
+    if (ctor_ == nullptr) {
+      initialized_.dec();
+    }
+  }
+
+private:
+  inline StatsStructType& instantiate() override { return *internal_stats_.get(ctor_); }
+
+  // In order to preserve stat value continuity across a config reload, we need to automatically
+  // re-instantiate lazy stats when they are constructed, if there is already a live instantiation
+  // to the same stats. Consider the following alternate scenarios:
+
+  // Scenario 1: a cluster is instantiated but receives no requests, so its traffic-related stats
+  // are never instantiated. When this cluster gets reloaded on a config update, a new lazy-init
+  // block is created, but the stats should again not be instantiated.
+
+  // Scenario 2: a cluster is instantiated and receives traffic, so its traffic-related stats are
+  // instantiated. We must ensure that a new instance for the same cluster gets its lazy-stats
+  // instantiated before the previous cluster of the same name is destructed.
+
+  // To do that we keep an "initialized" gauge in the cluster's scope, which will be associated by
+  // name to the previous generation's cluster's lazy-init block. We use the value in this shared
+  // gauge to determine whether to instantiate the lazy block on construction.
+  // TODO(#26106): See #14610. The initialized_ gauge could be disabled in a
+  // corner case where a user disables stats with suffix "initialized". In which case, the
+  // initialized_ will be a NullGauge, which breaks the above scenario 2.
+  // TODO(#26106): Consider hiding this Gauge from being exported, through using the
+  // stats flags mask.
+  Gauge& initialized_;
+  // TODO(#26957): Clean up this ctor_ by moving its ownership to AtomicPtr, and drop
+  // the setter lambda when the nested object is created.
+  std::function<StatsStructType*()> ctor_;
+  Thread::AtomicPtr<StatsStructType, Thread::AtomicPtrAllocMode::DeleteOnDestruct>
+      internal_stats_{};
```

Comment:
nit: construction with `{}` not needed here -- AtomicPtr is not a Pod -- it has a ctor.

Reply:
done